### PR TITLE
Feature : Balance Check persistent VM

### PR DESCRIPTION
**Pull request by 1yam**
Based on : https://github.com/aleph-im/pyaleph/pull/462

Solutions:
    - Add check of the balance for VM persistent
    - Fix View who did not show Persistent VM

### CODE REVIEW

File: src/aleph/handlers/content/vm.py [hunk 1]
```diff
@@ -320,15 +320,19 @@ class VmMessageHandler(ContentHandler):
     """
 
     async def check_balance(self, session: DbSession, message: MessageDb) -> None:
-        if message.type != MessageType.instance:
+        if not (message.type == MessageType.instance) and not (message.type == MessageType.program):
```
Comment (odesenfans) [inline src/aleph/handlers/content/vm.py:323]: `_get_vm_content()` called below already performs this check, you can just remove this line.

File: src/aleph/handlers/content/vm.py [hunk 2]
```diff
@@ -320,15 +320,19 @@ class VmMessageHandler(ContentHandler):
...
             return
+
         content = _get_vm_content(message)
-        if isinstance(content, ProgramContent):
-            return
+
+        if message.type == MessageType.program:
```
Comment (odesenfans) [inline src/aleph/handlers/content/vm.py:328]: It's easier for mypy to figure out what's going on if you use `if isinstance(content, ProgramContent)`.

File: src/aleph/handlers/content/vm.py [hunk 3]
```diff
@@ -320,15 +320,19 @@ class VmMessageHandler(ContentHandler):
...
+            if not content.on.persistent:
+                return
 
         required_tokens = compute_cost(session=session, content=content)
+
         current_balance = (
-            get_total_balance(address=content.address, session=session) or 0
+                get_total_balance(address=content.address, session=session) or 0
```
Comment (odesenfans) [inline src/aleph/handlers/content/vm.py:335]: Formatting, please apply `black` to all your PRs when opening them. Whitespace changes are just noise in the Git history.

File: src/aleph/services/cost.py [hunk 4]
```diff
@@ -12,9 +13,8 @@
 
 
 def _get_file_from_ref(
-    session: DbSession, ref: str, use_latest: bool
+        session: DbSession, ref: str, use_latest: bool
```
Comment (odesenfans) [inline src/aleph/services/cost.py:16]: Same, please format your PRs with `black`. It seems like you have a different formatter running automatically.

File: src/aleph/services/cost.py [hunk 5]
```diff
@@ -46,12 +46,13 @@ def get_volume_size(session: DbSession, content: ExecutableContent) -> Decimal:
     total_volume_size: Decimal = Decimal(0)
 
     for volume in ref_volumes:
-        file = _get_file_from_ref(
-            session=session, ref=volume.ref, use_latest=volume.use_latest
-        )
-        if file is None:
-            raise RuntimeError(f"Could not find entry in file tags for {volume.ref}.")
-        total_volume_size += Decimal(file.size)
+        if hasattr(volume, "ref"):
```
Comment (odesenfans) [inline src/aleph/services/cost.py:49]: Why did you add this condition? `ref_volumes` is defined as "all the volumes having a `ref` field" so this seems unneeded.

File: src/aleph/services/cost.py [hunk 6]
```diff
@@ -77,7 +78,18 @@ def get_additional_storage_price(
 def compute_cost(session: DbSession, content: ExecutableContent) -> Decimal:
     is_microvm = isinstance(content, ProgramContent) and not content.on.persistent
     compute_unit_cost: Decimal = Decimal("200.0") if is_microvm else Decimal("2000.0")
+    cpu: Decimal = Decimal(content.resources.vcpus)
+    memory: Decimal = Decimal(math.ceil(content.resources.memory / 2000))
+    compute_unit_multiplier = cpu if cpu >= memory else memory
+    environment_internet: Decimal = Decimal(0)
```
Comment (odesenfans) [inline src/aleph/services/cost.py:84]: For clarity and similarity with the view, could you change this to:

1. Compute the number of compute units in a function
2. Instead of `environment_internet`, could you add another function to return a "compute unit multiplier" based on the content of the message? Also you can use integers here, conversion to decimal will be automatic.

File: src/aleph/services/cost.py [hunk 7]
```diff
@@ -77,7 +78,18 @@ def get_additional_storage_price(
 def compute_cost(session: DbSession, content: ExecutableContent) -> Decimal:
     is_microvm = isinstance(content, ProgramContent) and not content.on.persistent
     compute_unit_cost: Decimal = Decimal("200.0") if is_microvm else Decimal("2000.0")
+    cpu: Decimal = Decimal(content.resources.vcpus)
+    memory: Decimal = Decimal(math.ceil(content.resources.memory / 2000))
+    compute_unit_multiplier = cpu if cpu >= memory else memory
```
Comment (odesenfans) [inline src/aleph/services/cost.py:83]: Rename to `nb_compute_units`.

File: src/aleph/services/cost.py [hunk 8]
```diff
@@ -77,7 +78,18 @@ def get_additional_storage_price(
...
+    environment_internet: Decimal = Decimal(0)
+
+    if isinstance(content, ProgramContent) and content.environment.internet:
+        environment_internet = Decimal(2)
 
-    return (compute_unit_cost * content.resources.vcpus) + get_additional_storage_price(
+    compute_units_required = compute_unit_multiplier
+    compute_unit_multiplier = environment_internet or 1
```
Comment (odesenfans) [inline src/aleph/services/cost.py:90]: As mentioned in the comment above, this can be simplified to

```
compute_unit_multiplier = 1
if content.environment.internet:
  compute_unit_multiplier += 1
```

File: tests/message_processing/test_process_programs.py [hunk 9]
```diff
@@ -45,7 +45,7 @@ def fixture_program_message(session_factory: DbSessionFactory) -> PendingMessage
...
...acanet","TRUSTED_RPC_ENDPOINT":"https://rpc.tzkt.io/ithacanet","WELL_CONTRACT":"KT1ReVgfaUqHzWWiNRfPXQxf7TaBLVbxrztw","PORT":"8080","CONCURRENT_JOB":"5","BATCH_SIZE":"10","UNTIL_BLOCK":"201396","PUBSUB":"{\\"namespace\\": \\"tznms\\",\\"uuid\\": \\"tz_uid_1\\",\\"hook_url\\": \\"_domain_or_ip_addess\\",\\"pubsub_server\\": \\"domain_or_ip_address\\",\\"secret_shared_key\\": \\"112secret_key\\",\\"channel\\": \\"storage\\",\\"running_mode\\": \\"readonly\\"}"},"on":{"http":false, "persistent": true},"environment":{"reproducible":false,"internet":true,"aleph_api":true,"shared_cache":false},"resources":{"vcpus":4,"memory":4095,"seconds":300},"runtime":{"ref":"bd79839bf96e595a06da5ac0b6ba51dea6f7e2591bb913deccded04d831d29f4","use_latest":true,"comment":"Aleph Alpine Linux with Python 3.8"},"volumes":[{"comment":"Extra lib","mount":"/opt/extra_lib","ref":"5f31b0706f59404fad3d0bff97ef89ddf24da4761608ea0646329362c662ba51","use_latest":true},{"comment":"Python Virtual Environment","mount":"/opt/packages","ref":"1000ebe0b61e41d5e23c10f6eb140e837188158598049829f2820f830139fc7d","use_latest":true},{"comment":"Data storage","mount":"/data","persistence":"host","name":"data","size_mib":128}]}',
```
Comment (odesenfans) [inline tests/message_processing/test_process_programs.py:48]: What did you change here? If it's just formatting, please revert.

**Review by odesenfans — CHANGES_REQUESTED** [review]

File: tests/message_processing/test_process_programs.py (hunk 9 above)
Comment (1yam) [reply tests/message_processing/test_process_programs.py:48]: `"on":{"http":true}`
vs 
`"on":{"http":false, "persistent": true`

**Review by odesenfans — APPROVED** [review]
